dataprep/clean/clean_email.py

```diff
@@ -0,0 +1,515 @@
+    """
+    domains_map = set(DOMAINS)
+    if domain_part.lower() not in domains_map:
+        for i, curr_c in enumerate(domain_part):
+            # two neighbor chars in reverse order
+            if (domain_part[0:i] + domain_part[i + 1 :]).lower() in domains_map:
+                domain_part = (domain_part[0:i] + domain_part[i + 1 :]).lower()
+                break
+
+            # missing single char
+            for new_c in "abcdefghijklmnopqrstuvwxyz":
+                if (
+                    domain_part[0 : i + 1] + new_c + domain_part[i + 1 :]
+                ).lower() in domains_map:
+                    domain_part = (
+                        domain_part[0 : i + 1] + new_c + domain_part[i + 1 :]
+                    ).lower()
+                    break
+
+            # redundant single char
+            if i < len(domain_part) - 1:
+                if (
+                    domain_part[0:i]
+                    + domain_part[i + 1]
+                    + curr_c
+                    + domain_part[i + 2 :]
+                ).lower() in domains_map:
+                    domain_part = (
+                        domain_part[0:i]
+                        + domain_part[i + 1]
+                        + curr_c
+                        + domain_part[i + 2 :]
+                    ).lower()
+                    break
+
+            # misspelled single char
+            if curr_c in NEARBYKEYS:
+                for c_p in NEARBYKEYS[curr_c]:
+                    if (
+                        domain_part[0:i] + c_p + domain_part[i + 1 :]
+                    ).lower() in domains_map:
+                        domain_part = (
+                            domain_part[0:i] + c_p + domain_part[i + 1 :]
+                        ).lower()
+                        break
+    return domain_part
+
+
+def format_email(
+    row: pd.Series,
+    col: str,
+    split: bool,
+    pre_clean: bool,
+    fix_domain: bool,
+    error: str,
+) -> pd.Series:
+    """
+    Function to transform an email address into clean format
+    """
+    # pylint: disable=too-many-nested-blocks, too-many-locals,too-many-branches,too-many-statements,too-many-return-statements, too-many-arguments
+
+    # pre-cleaning email text by removing all whitespaces
+    if pre_clean:
+        row[col] = re.sub(
+            r"(\s|\u180B|\u200B|\u200C|\u200D|\u2060|\uFEFF)+", "", str(row[col])
+        )
+
+    valid_type = check_email(row[col], True)
+
+    if valid_type == "null":
+        return not_email(row, col, split, "null", error)
+
+    elif valid_type == "bad_format":
+        return not_email(row, col, split, "bad_format", error)
+
+    elif valid_type == "overflow":
+        return not_email(row, col, split, "overflow", error)
+
+    elif valid_type == "unknown":
+        return not_email(row, col, split, "unknown", error)
+
+    elif valid_type != "valid":
+        raise ValueError("invalid return type for instance check")
```

Comment:
In clean_lat_long we output the value that's unable to be parsed. Do you think we should print the value here too?

Reply:
I specified a few possible return type for validation function. If the returned type is not any of them, an exception will be thrown. For any other specified types, I'll report values.